cmd/ceremony/key.go

```diff
@@ -0,0 +1,158 @@
+package main
+
+import (
+	"bytes"
+	"crypto"
+	"crypto/x509"
+	"encoding/pem"
+	"errors"
+	"fmt"
+	"io/ioutil"
+	"log"
+
+	"github.com/letsencrypt/boulder/pkcs11helpers"
+	"github.com/miekg/pkcs11"
+)
+
+type generateArgs struct {
+	mechanism    []*pkcs11.Mechanism
+	privateAttrs []*pkcs11.Attribute
+	publicAttrs  []*pkcs11.Attribute
+}
+
+func getRandomBytes(ctx pkcs11helpers.PKCtx, session pkcs11.SessionHandle) ([]byte, error) {
+	r, err := ctx.GenerateRandom(session, 4)
+	if err != nil {
+		return nil, err
+	}
+	return r, nil
+}
+
+const (
+	rsaModLen = 2048
+	rsaExp    = 65537
+)
+
+type keyInfo struct {
+	key interface{}
+	der []byte
+	id  []byte
+}
+
+func generateKey(ctx pkcs11helpers.PKCtx, session pkcs11.SessionHandle, config ceremonyConfig) (*keyInfo, error) {
+	var pubKey interface{}
+	var keyID []byte
+	var err error
+	switch config.KeyType {
+	case "RSA":
+		pubKey, keyID, err = rsaGenerate(ctx, session, config.KeyLabel, rsaModLen, rsaExp)
+		if err != nil {
+			return nil, fmt.Errorf("failed to generate RSA key pair: %s", err)
+		}
+	case "ECDSA":
+		pubKey, keyID, err = ecGenerate(ctx, session, config.KeyLabel, config.ECDSACurve)
+		if err != nil {
+			return nil, fmt.Errorf("failed to generate ECDSA key pair: %s", err)
+		}
+	}
+
+	der, err := x509.MarshalPKIXPublicKey(pubKey)
+	if err != nil {
+		return nil, fmt.Errorf("Failed to marshal public key: %s", err)
+	}
+
+	pemBytes := pem.EncodeToMemory(&pem.Block{Type: "PUBLIC KEY", Bytes: der})
+	log.Printf("Public key PEM:\n%s\n", pemBytes)
+	if err := ioutil.WriteFile(config.PublicKeyPath, pemBytes, 0644); err != nil {
+		return nil, fmt.Errorf("Failed to write public key to %q: %s", config.PublicKeyPath, err)
+	}
+	log.Printf("Public key written to %q\n", config.PublicKeyPath)
+	return &keyInfo{key: pubKey, der: der, id: keyID}, nil
+}
+
+// findObject looks up a PKCS#11 object handle based on the provided template.
+// In the case where zero or more than one objects are found to match the
+// template an error is returned.
+func findObject(ctx pkcs11helpers.PKCtx, session pkcs11.SessionHandle, tmpl []*pkcs11.Attribute) (pkcs11.ObjectHandle, error) {
+	if err := ctx.FindObjectsInit(session, tmpl); err != nil {
+		return 0, err
+	}
+	handles, more, err := ctx.FindObjects(session, 1)
+	if err != nil {
+		return 0, err
+	}
+	if len(handles) == 0 {
+		return 0, errors.New("no objects found matching provided template")
+	}
+	if more {
+		return 0, errors.New("more than one object matches provided template")
+	}
+	if err := ctx.FindObjectsFinal(session); err != nil {
+		return 0, err
+	}
+	return handles[0], nil
+}
+
+// getKey constructs a x509Signer for the private key object associated with the
+// given label and ID
+func getKey(ctx pkcs11helpers.PKCtx, session pkcs11.SessionHandle, label string, id []byte) (*x509Signer, error) {
```

Comment:
In the `pkcs11key` package, we modified the interfaces so that you have to provide a public key when searching for a private key, and the lookup functions will return an error unless there's a match. This avoid a whole category of errors where you sign something with the wrong private key and get an invalid signature. I think when you first landed this code you had an argument for why that's not appropriate here, but I'm afraid I don't recall it. If it's still relevant, could you comment this code with the reason? And if it's not still relevant, could you change to an approach that makes it impossible to get a private key handle unless it matches a provided public key?

Also, I think this should return a `crypto.Signer`. The caller doesn't need to know the detail that this is implemented under the hood as an `*x509Signer`.